### PR TITLE
Add tests copy phrase

### PR DESCRIPTION
**Pull request by nik-sm**
# Overview

Added a smoke test for CopyPhraseWrapper to prepare for refactoring PcaRdaKde model.

Simplified code in query_mechanisms.py and add test.
Note - in the case of a tie, the new code takes items from the rear of the list, rather than from the front.
This behavior was not tested and presumably only occurs in the case of a uniform alphabet distribution.

## Ticket

https://www.pivotaltracker.com/story/show/177952877 ("add test for CopyPhraseWrapper")

https://www.pivotaltracker.com/story/show/177856966 ("Separate integration from unit tests")

## Contributions

- `bcipy/helpers/tests/test_copy_phrase_wrapper.py` - renamed the file, and added a smoke test for uncovered methods `CopyPhraseWrapper.initialize_series()` and `CopyPhraseWrapper.evaluate_inquiry()`
- `bcipy/signal/tests/model/test_offline_analysis.py` - mark integration tests as "slow"  (test can be skipped with `pytest -k "not slow"` or `pytest -k "MyKeyword and not slow"`)
- `bcipy/tasks/rsvp/query_mechanisms.py` - make base class into proper abstract base class. Shorten the `best_selection` method.
- `bcipy/tasks/tests/test_query_mechanisms.py` - add test for `best_selection` method. As noted, the method breaks ties differently than before.

## Test

`pytest  --disable-warnings -k "TestCopyPhraseWrapper or TestQueryMechanisms"`

## Documentation

Updated docstrings

### CODE REVIEW

File: bcipy/tasks/rsvp/query_mechanisms.py [hunk 1]
```diff
@@ -183,25 +168,8 @@ def do_series(self):
...
-        val = np.delete(val, idx_q)
-        list_el.remove(q)
-        query.append(q)
-
-    return query
+def best_selection(list_el: List[Any], val: List[float], len_query: int):
+    """Return the top `len_query` items from `list_el` according to the values in `val`"""
+    # numpy version: return list_el[(-val).argsort()][:len_query]
+    sorted_items = reversed(sorted(zip(val, list_el)))
+    return [el for (value, el) in sorted_items][:len_query]
```
Comment (lawhead) [inline bcipy/tasks/rsvp/query_mechanisms.py:174]: This is a nice simplification.

File: Makefile [hunk 2]
```diff
@@ -10,6 +10,12 @@ test-all:
 	flake8 bcipy
 	coverage report
 
+unit-test:
+	pytest -k "not slow"
+
+integration-test:
+	pytest -k "slow"
+
```
Comment (lawhead) [inline Makefile:18]: Thanks for separating these out.

File: bcipy/helpers/tests/test_copy_phrase_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+            alp=alp,
+            task_list=[("HELLO_WORLD", "HE")],
+            is_txt_stim=True,
+            device_name=self.device_spec.name,
+            evidence_names=["LM", "ERP"],
+            device_channels=self.device_spec.channels,
+            stimuli_timing=[0.5, 0.25],
+            notch_filter_frequency=4.0,
+            filter_low=1.0,
+            filter_high=4.0,
```
Comment (lawhead) [inline bcipy/helpers/tests/test_copy_phrase_wrapper.py:184]: You may want to document how you arrived at these filter values.

File: bcipy/helpers/tests/test_copy_phrase_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+    def test_init_series_evaluate_inquiry(self):
+        # TODO - it would be just as easy to store a snippet of real data, which would be a much better test
+        # (since this fake data likely does not exercise the dynamic range of the model)
+        alp = alphabet()
+
+        # Create fake data to provide as the user's response
+        # Create test items that resemble the fake training data, testing inference
+        # TODO - is this duration calculated precisely, or is it just large enough by coincidence?
+        # int(self.num_trial * self.device_spec.sample_rate * (self.params["time_flash"] + self.params["static_trigger_offset"]))
```
Comment (lawhead) [inline bcipy/helpers/tests/test_copy_phrase_wrapper.py]: I think we forgot to add self.params['trial_length'] here. See `process_data_for_decision` in the `task.py` helper module . The default value used if the `buf_len` param is not provided is the `trial_length` parameter.

File: bcipy/helpers/tests/test_copy_phrase_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        self.assertEqual(expected_letters, letters)
+        self.assertEqual(expected_time, timings)
+        self.assertEqual(["nontarget"] * (len(letters)), labels)
+
+        # Test it throws an exception when letter is outside alphabet
+        with self.assertRaises(Exception):
+            cp.letter_info([("A", 0.0), ("*", 1.0)], ["nontarget", "nontarget"])
+
+    def test_init_series_evaluate_inquiry(self):
+        # TODO - it would be just as easy to store a snippet of real data, which would be a much better test
```
Comment (lawhead) [inline bcipy/helpers/tests/test_copy_phrase_wrapper.py]: I'm fine with that approach.

**Review by lawhead — COMMENTED** [review]
This seems like a good integration test that meets the acceptance criteria of the ticket.

**Review by lawhead — APPROVED** [review]

File: bcipy/helpers/tests/test_copy_phrase_wrapper.py (hunk 4 above)
Comment (nik-sm) [reply bcipy/helpers/tests/test_copy_phrase_wrapper.py]: Ok great - `num_trial * sample_rate * trial_length` works!

**Pull request merged**